cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -674,6 +684,53 @@ Status CopyFiles(const std::shared_ptr<FileSystem>& source_fs,
   return CopyFiles(sources, destinations, io_context, chunk_size, use_threads);
 }
 
+struct Registry {
+  std::shared_mutex mutex;
+  std::unordered_map<std::string, FileSystem::Factory*> scheme_to_factory;
+  std::vector<void (*)()> finalizers;
+  bool finalized = false;
+};
```

Comment:
Could these members be made `private` and all functions that manipulate them (including the mutex) become part of the class itself?

Reply:
Certainly. Since this is a private detail of the registry I was being more informal